src/Functions/FunctionsStringArray.h

```diff
@@ -493,11 +510,16 @@ class SplitByRegexpImpl
     {
         pos = pos_;
         end = end_;
+        curr_tokens = 0;
     }
 
     /// Get the next token, if any, or return false.
     bool get(Pos & token_begin, Pos & token_end)
     {
+        /// Returned tokens exceeds max_tokens limitation.
+        if (max_tokens && *max_tokens && curr_tokens >= *max_tokens)
+            return false;
+
```

Comment:
Sort of unrelated: the check `if (!re)` (l. 523) could be done at the end of the `init()` method (basically throw an exception ... More importantly, the only reason for `!re` is an out-of-memory/bad_alloc, an invalid regexp will throw an exception already (see the `OptimizedRegularExpressionImpl` ctor in Common/OptimizedRegularExpression.cpp). But a bad_alloc is also an exception, so we might remove `if (!re)` altogether.